### ISSUE TITLE
Pade Spectral Extrapolation

### ISSUE THREAD

**Issue by Arjun-Khurana** [NanoComp/meep#2192]
The following are some initial results for using Pade Approximant-based spectral extrapolation to reduce simulation time for resonant structures (introduced in [this issue](https://github.com/NanoComp/meep/issues/375#issue-330507605) and expanded upon [here](https://github.com/NanoComp/meep/issues/1497#issue-803911576).

The structure used is a simple ring resonator (as in [this issue](https://github.com/NanoComp/meep/issues/2183#issue-1324726574)). Full code is available [here](https://github.com/Arjun-Khurana/series-extrapolation.git).

The time series field data of the ring resonator are sampled (currently with a sampling rate of 1, so every point is considered. Future investigation will decimate the time series data to potentially reduce spatial complexity). 

These data are used as input to `scipy.interpolate.pade` (further work will investigate implementing a more robust or faster Pade algorithm, as this operation is rather slow).  These Pade coefficients are then used with a Fourier basis $e^{i \omega n \Delta t}$ with the same frequencies $\omega$ as the Meep DTFT calculation to construct a spectral extrapolation of the field data which is comparable to the Meep DTFT. 

Comparing the Meep DTFT calculation at 200 simulation time steps (about 5 periods of the ring resonator) with the Pade extrapolation at 200 time steps:

![extrapolation_t=200 0](https://user-images.githubusercontent.com/35874188/184214033-8b930149-b38c-46f6-b813-2110fbaa9633.png)

Clearly there is significantly less spectral leakage in the Pade extrapolation. At very low numbers of time steps (<3 periods of the ring): 

![extrapolation_t=100 0](https://user-images.githubusercontent.com/35874188/184214511-4cff6316-f3a8-473c-a792-073763091a12.png)

There is seemingly not much benefit to using the Pade extrapolation at these time.

Examining the long term convergence behavior of the Meep DTFT, I ran the simulation for 100 periods (4000 time steps) and calculated the DTFT at various points:

![dft-extension](https://user-images.githubusercontent.com/35874188/184214755-019c72e1-19bf-4597-b94b-f48155cfba58.png)

The DTFT has a small amount spectral leakage after approximately 75 periods, and minimal after 100 periods. Comparing the Pade extrapolation after 5 periods with the Meep DTFT after 100 periods:

![pade-winning](https://user-images.githubusercontent.com/35874188/184215363-c1e1aba5-039b-428a-a5cb-61c61af73d2a.png)

For this example, it seems Pade extrapolation can potentially save a factor of 20 in simulation time.

One unfortunate result however is that these results are all normalized. The raw Pade computation has a wildly different magnitude than the Meep DTFT, and I am as yet unsure of the cause: 

![unnormalized](https://user-images.githubusercontent.com/35874188/184216280-d4091c95-f1ee-4765-964d-357c0863026e.png)

**Comment by smartalecH** [comment]
Linking #1497 and #375.

**Comment by stevengj** [comment]
I would run meep for about 10Q, about 1000 periods, in order to get an "exact" reference spectrum.

**Comment by stevengj** [comment]
Other things to try:

* a higher-Q example
* an example with a more wiggly specturm, for example the transmission spectrum in homework problem 2(a) here: https://github.com/mitmath/18369/blob/master/psets/pset5sol.ipynb
* an example that has some other singularity than a simple pole, e.g. the van Hove 1/sqrt(ω) singularities that appear in the LDOS in problem 2(d) here: https://github.com/mitmath/18369/blob/master/psets/pset3sol-p2.ipynb

**Comment by Arjun-Khurana** [comment]
> I would run meep for about 10Q, about 1000 periods, in order to get an "exact" reference spectrum.

![dft-extension-2](https://user-images.githubusercontent.com/35874188/184244522-eac393b0-38b8-468f-ac6b-396b9733ed40.png)

![pade-winning-2](https://user-images.githubusercontent.com/35874188/184244549-d2ddfb48-548c-48bc-affe-b116d854584e.png)

**Comment by Arjun-Khurana** [comment]
Updated results for different structures:

### Ring Resonator ###
```
harminv0:, frequency, imag. freq., Q, |amp|, amplitude, error
harminv0:, 0.14740026347983531, -0.00021470863462223294, 343.25648742347346, 0.011130041872620238, 0.0079490599548909+0.007790396518780653i, 9.545963689117064e-08+0.0i
harminv0:, 0.17571948081064348, -5.359646259508504e-05, 1639.2824479684737, 0.05227488897978794, -0.012419114315467016-0.05077823960584412i, 2.723932613912651e-09+0.0i
harminv0:, 0.20341270591683974, -1.2491877354800731e-05, 8141.798872155376, 0.010167981927680616, -0.004776961539371981+0.00897599548423463i, 2.3699126777806563e-09+0.0i
```

![ring](https://user-images.githubusercontent.com/35874188/185468289-60d93601-9f66-48a8-8225-5caf778c1f7d.png)

### Waveguide FP Cavity, N=3 ###
```
harminv0:, frequency, imag. freq., Q, |amp|, amplitude, error
harminv0:, 0.23485262845690733, -0.0003099455770774247, 378.8610740495273, 0.0015085991075502007, 0.0014033853927133027+0.0005535166725765278i, 4.3091088742332866e-08+0.0i
```

![cavity_n=3](https://user-images.githubusercontent.com/35874188/185468479-f7b7b6dc-dd6d-4007-98e4-aa2089880a81.png)

### Waveguide FP Cavity, N=5 ###
```
harminv0:, frequency, imag. freq., Q, |amp|, amplitude, error
harminv0:, 0.19151697744743398, -0.0015832609343938593, 60.48181107959787, 0.0011836969791242302, 0.00105812124001351-0.0005305826795327003i, 1.2216533654819396e-07+0.0i
harminv0:, 0.23493744725885676, -2.8741329404868164e-05, 4087.1012601641073, 4.727930035768515e-05, 4.227213214962697e-05+2.117543590665091e-05i, 2.116446030465839e-09+0.0i
harminv0:, 0.3119427422105216, -0.0011378655591419446, 137.0736374364627, 0.0007730036384326723, -7.401770943987447e-05-0.0007694517552903651i, 4.083006688516505e-08+0.0i
harminv0:, 0.3157812041160235, -0.002917290344141031, 54.12234760078403, 0.0017586337254288165, -6.0622703221663386e-05+0.001757588537761252i, 1.8502203890278337e-07+0.0i
harminv0:, 0.3346881619665811, -0.0028699117290561437, 58.30983555662412, 7.270714447461183e-05, -6.373477863941507e-05-3.498866744299987e-05i, 6.297137516590531e-06+0.0i
```

![cavity_n=5](https://user-images.githubusercontent.com/35874188/185468580-97adfd65-33e1-4114-ad45-112432c752c1.png)

### Waveguide FP Cavity, N=7 ###
```
harminv0:, frequency, imag. freq., Q, |amp|, amplitude, error
harminv0:, 0.19360822136007844, -0.0010222824980862188, 94.69408980517909, 0.0008953076433468013, -0.0007240832719189372+0.0005265730638404985i, 1.5959121885350877e-07+0.0i
harminv0:, 0.19520892993049999, -0.000691346513398236, 141.1802375128591, 0.0006708893143463233, 0.00048647749422788423-0.00046198714236852807i, 7.968146346648642e-08+0.0i
harminv0:, 0.23494102233357728, -1.1912751411660965e-05, 9860.905101385812, 7.533688756179789e-07, 7.485805644764937e-07+8.480448831253985e-08i, 1.4654922816149303e-07+0.0i
harminv0:, 0.3101593140316803, -0.00041095844610105156, 377.36091930254975, 0.00041504650546023456, 0.00024666471899754147-0.00033379652199598547i, 2.37044691043493e-08+0.0i
harminv0:, 0.3114400061646693, -0.0005485874680252668, 283.8562893951535, 0.0005678697490809493, -0.00020152209217609335+0.0005309095010321727i, 3.6627476285440714e-08+0.0i
harminv0:, 0.3171722067778276, -0.0030676163098192094, 51.69685103097529, 0.0009607063299982731, -0.0006574116770845204-0.0007005473141277963i, 1.2741879574848883e-06+0.0i
harminv0:, 0.32508052924138975, -0.0017038462192598709, 95.3960884400121, 0.00041927835043618623, 2.907366619336383e-05+0.00041826912039805934i, 7.714247462218277e-07+0.0i
```

![cavity_n=7](https://user-images.githubusercontent.com/35874188/185468683-8a519b54-5ea2-406f-85f9-fbe08fdb3c47.png)

![cavity_n=7_narrow](https://user-images.githubusercontent.com/35874188/185468702-5b0f2447-a26e-44b3-95bc-9975a9587554.png)

### Metallic Waveguide, Off Center ###
```
harminv0:, frequency, imag. freq., Q, |amp|, amplitude, error
harminv0:, 1.4897661300903537, -4.124677755628634e-05, 18059.18205437241, 0.008869036490940875, -0.006897026495143153+0.005575915512535515i, 7.799952280859097e-09+0.0i
harminv0:, 1.4949597550433054, -0.0011026795857930496, 677.8758645323642, 0.012162854428269377, 0.01210697339964973+0.0011645698533990479i, 8.921420708377969e-07+0.0i
harminv0:, 1.4989446604445396, -0.0023468079291866376, 319.3581890112429, 0.017719069059104406, 0.009783808112404316+0.014773033105661224i, 3.786321169915711e-06+0.0i
harminv0:, 1.5150052816498942, -0.005711401091247483, 132.62991492329135, 0.028674222543938074, 0.017457399200629626-0.02274753286950351i, 2.8186171930565817e-05+0.0i
harminv0:, 1.522442789343261, -0.008665027366976904, 87.8498546435877, 0.05213958155528552, 0.018565008077628162-0.04872244287633648i, 3.571071793530682e-05+0.0i
harminv0:, 2.4514832882080264, -1.3984209845804794e-05, 87651.83429164077, 0.014432157292666463, -0.007561128291159144-0.012292945256727477i, 1.3514589761847738e-09+0.0i
harminv0:, 2.452523168013328, -0.00013071331744846765, 9381.305653803061, 0.014846883184167551, 0.009292128471242226+0.011579563409656117i, 5.008524475087916e-09+0.0i
harminv0:, 2.4546091064960702, -0.000387088113792017, 3170.607697624804, 0.018016053432476437, 0.005156510421087679+0.017262345772202i, 2.460726796724591e-08+0.0i
harminv0:, 2.4620624683197825, -0.0013614868244751456, 904.1815256893543, 0.021513627415897046, 0.005415455343898914-0.020820879136296665i, 2.7762483862753473e-07+0.0i
harminv0:, 2.4667069702583784, -0.0020656041943948817, 597.0909085467363, 0.026633131277857238, 0.026531758227657746+0.0023215268709823244i, 6.643940307042968e-07+0.0i
harminv0:, 2.4723626156460163, -0.002318672984149341, 533.1417221288452, 0.017085469645015814, 0.01523862096502796+0.007726428934182149i, 4.623276266094154e-07+0.0i
harminv0:, 2.4811660825010438, -0.003496761844952604, 354.78053589529975, 0.021218716675226252, -0.008964370088848927-0.019232108731329366i, 1.920248022429829e-06+0.0i
harminv0:, 2.4890669946616897, -0.00439704905098327, 283.03834751457725, 0.02873056255390957, 0.009393138408193731-0.027151688262585558i, 1.7201305893888018e-06+0.0i
harminv0:, 2.496157161962382, -0.007911656022939908, 157.75187613849457, 0.08075144879936019, 0.05832136970629335-0.05585171724107987i, 1.0807090951433839e-05+0.0i
harminv0:, 2.5066600508136774, -0.003119239064285773, 401.8063378844672, 0.003913057321887651, 0.003908836627925909+0.00018169705711256346i, 1.2926531174772938e-05+0.0i
harminv0:, 2.52109836775153, -0.005304440542341236, 237.6403644859016, 0.008280658510955597, -0.0011148538209460475-0.00820526698730659i, 4.340631116094957e-06+0.0i
harminv0:, 2.5313758942421725, -0.01933587677371473, 65.45800647848912, 0.35676130417812446, -0.09390455704117887-0.344181002273188i, 1.2591663487118931e-05+0.0i
```

![ldos_x=w](https://user-images.githubusercontent.com/35874188/185468957-2e041dfe-a118-4083-a472-ebfd2c13ac18.png)

### Metallic Waveguide, Center ###
```
harminv0:, frequency, imag. freq., Q, |amp|, amplitude, error
harminv0:, 1.4897517584191544, -4.06746496694492e-05, 18313.025072445915, 0.012244623740498886, -0.007294673713522692+0.009834558757749632i, 5.245177767355653e-09+0.0i
harminv0:, 1.4930519536010007, 0.002015419292024706, -370.4072793957105, 0.0065618846541944135, 0.006331871149291941+0.001722131807881294i, 1.4840872833635142e-06+0.0i
harminv0:, 1.4947424490880992, -0.00048149022124109597, 1552.2043679674637, 0.008515567772241227, -0.0059296336222065685+0.006111819662754362i, 6.883351595309382e-07+0.0i
harminv0:, 1.4994405143337315, -0.0016071245380756827, 466.49792185026047, 0.00889343664731406, -0.0034553397240866352+0.008194744827689122i, 1.1867649640461599e-06+0.0i
harminv0:, 1.505521835287647, -0.0016913252501232193, 445.0716487493946, 0.004785327803268485, 0.001095958440393791+0.004658136674858736i, 2.262163318530374e-06+0.0i
harminv0:, 1.525176731651881, -0.0013820865457606318, 551.7660005916994, 0.001319751504651585, -0.001300332986447676+0.0002255618726340761i, 5.251048033138766e-06+0.0i
harminv0:, 1.5938753709125755, -0.015410640798545415, 51.71346836735753, 0.09461388543326797, -0.08988541324533246-0.02953641485514735i, 0.00047892349005565553+0.0i
harminv0:, 2.4490508168133784, 0.0015114725879892016, -810.1538976871194, 0.041713327932387564, -0.03420913715273641-0.023869157137615362i, 0.00017660338246940196+0.0i
harminv0:, 2.4514756186443565, -1.725581075209251e-05, 71033.3363602484, 0.02120802374549322, -0.013529490910429374-0.016331967024642596i, 3.611533373287346e-07+0.0i
harminv0:, 2.4524898155553188, -0.00013582326585243976, 9028.239013998298, 0.021984146032977443, -0.018033951668924027-0.012572957647355304i, 6.828481749640708e-07+0.0i
harminv0:, 2.457578982077911, -0.0006958086607152413, 1765.9876348418086, 0.01873851399508163, -0.015404290767865546-0.010669570407608844i, 2.312515054219684e-06+0.0i
harminv0:, 2.4671410618318643, -0.001456339548738092, 847.0349733925804, 0.01401669408704458, -0.006055111882518763-0.012641334313274473i, 1.593046433623244e-06+0.0i
harminv0:, 2.4728123576309864, -0.0023438439438695346, 527.5121588403479, 0.015990008048561502, -0.01410408321719895-0.007533604316354063i, 2.9262711983135976e-06+0.0i
harminv0:, 2.4729617706924065, 0.0029458592785952943, -419.73521760883557, 0.0007699231352635401, 0.000246282420862679-0.0007294700839568802i, 7.190620903478571e-05+0.0i
harminv0:, 2.4889236422451075, -0.0027512609193794035, 452.32417338420396, 0.0066967629144457225, -0.0029708279291429914-0.006001734328318735i, 5.6921781987199545e-06+0.0i
harminv0:, 2.49697792428688, 0.000339546266600563, -3676.933263448728, 2.0165442327968526e-06, -1.9395951767922794e-06-5.517437747635888e-07i, 2.08665777046642e-05+0.0i
harminv0:, 2.50833822255578, 0.0018657880508660127, -672.1927020037258, 0.0025198169861706614, 0.0012927389721016983-0.0021629386477206515i, 1.8579894050576637e-05+0.0i
harminv0:, 2.5211303679924466, 0.0037296149332288263, -337.98802465243205, 0.002235654116928981, 0.0022014226727219695-0.0003897275286237337i, 3.6526192658050174e-05+0.0i
harminv0:, 2.5367741897326077, -0.0004528815096397035, 2800.704086760768, 0.0012970392102914447, 0.0012857151811499607+0.0001710180867451693i, 1.3716274049755827e-05+0.0i
```

![ldos_x=0](https://user-images.githubusercontent.com/35874188/185469056-0c12296b-596e-486b-a7a9-8fd8d246846c.png)

The Pade Approximant for the DFT in the case of the metallic waveguide measured in the center does not converge to the long-term DTFT, even though the approximant works nicely when measuring off center for the same experiment. One discrepancy between these is the amplitude of the time domain field:

![ldos_amp](https://user-images.githubusercontent.com/35874188/185469271-c8fed5b7-16ad-4bbb-9bad-1dd76039f474.png)

While the Gaussian pulse is turned on, the amplitude of the field in the center of the waveguide is much larger than the amplitude when the pulse reaches other parts of the waveguide. Perhaps this is causing some failure of the Pade algorithm.

**Comment by stevengj** [comment]
Might be nice to plot the relative error |log(pade) - log(meep)| / |log(meep)|, where |...| is L2 norm, as a function of the number of timesteps used for pade.

**Comment by stevengj** [comment]
In general, Padé is going to have a hard time with lossless resonances where the DTFT doesn't even necessarily converge with time, and it will especially have a hard time with singularities that aren't simple poles — the van Hove singularity in the metallic waveguide is of this form, since it looks like 1/√(ω-ω₀), so I proposed it as an intentionally difficult test case.   So if these are the only cases where we fail I think it is still pretty useful!

I would try my photonic-crystal example, which is different from your holey-waveguide cavity example in that it has lots of low-loss (but not lossless) resonances (as well as exponentially small transmission in the gap).

**Comment by Arjun-Khurana** [comment]
> Might be nice to plot the relative error |log(pade) - log(meep)| / |log(meep)|, where |...| is L2 norm, as a function of the number of timesteps used for pade.

Error analysis for ring resonator: 

![ringnorms](https://user-images.githubusercontent.com/35874188/186746636-7d6537db-63ee-4c37-98ae-c04f165feb5a.png)

There is some interesting behavior as the Pade approximant does not improve referenced to the meep ground truth after a certain point. Does this represent the fundamental limit of the approximation?

Error analysis for varying resolutions and sampling rates of the input E-field data:

![ringnormsexp](https://user-images.githubusercontent.com/35874188/186748117-95d5aad7-79b6-4afe-bf9b-090001e4f147.png)

It seems we can decimate the input electric field data at a considerable rate and still construct a well-behaved Pade approximant. There should be a way to find the optimal decimation factor from the expected spectral content of the signal. 

Now trying the photonic crystal example:

![rodsnorms](https://user-images.githubusercontent.com/35874188/186747363-c983a261-ee6a-462e-9146-11722c73b976.png)

This example also performs well with the Pade approximant.

Returning to the metallic waveguide example:

![ldosnorms](https://user-images.githubusercontent.com/35874188/186747834-bd277691-79dd-4fec-a0e5-5e872ae609cd.png)


When the meep DTFT does not converge nicely with longer simulation time, neither does the Pade approximant.

**Comment by smartalecH** [comment]
@Arjun-Khurana any updates on this? Any chance we could wrap this functionality into the public API somehow?

**Comment by Arjun-Khurana** [comment]
> @Arjun-Khurana any updates on this? Any chance we could wrap this functionality into the public API somehow?

Will be pushing forward on this over the next 2-3 weeks (finals week over). Goals are to figure out the noise floor for rational function approximation and integrate with the adjoint source formulation.

**Comment by stevengj** [comment]
Closed by #2440
